nanoframework/nf-interpreter
# Update for Smart Config

## Description
The enables the Smart config for ESP32. This allows the WiFI  SSID & Password to be configured from a mobile phone.

When the nanoCLR starts and there is no SSID setup for the Wireless Station the nanoCLR will now start the smart config task.  This task watches for special wifi packets which allows the SSID & Password to be configured, connected and stored.  

A managed code application can monitor the state/IP address of the wireless station to see if it has been configured/connected before starting.

## Motivation and Context
To allow the SSID to be provisioned on site without having to use Visual studio
 
## How Has This Been Tested?

- Erased flash & load latest firmware
- From phone install Smart config app ( for example, Android :- "Esp8266 Smartconfig" ). The are others - that work.
- Connect phone to AP you want to configure on device.
- Run Smart config app
- Enter Password and configure device. ("Esp8266 Smartconfig" press confirm)
- Confirm device configured

When configured the device will connect to AP and shutdown Smart config.

Further testing was done to confirm Network functions were working (Dgrams and Stream sockets )

To restart Smart config you need to either reload firmware or from managed code/VS extension reset the SSID back to nothing and reset device.

No testing was done with STM32 targets so should be done as some common code was changed. 

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)

## Checklist:




Signed-off-by: adriansoundy <adriansoundy@gmail.com>
